### PR TITLE
[#1075] static string rust

### LINKED ISSUE #1075: Implement static string in Rust
## Brief feature description

Implement a compile-time fixed capacity string in Rust that is memory layout compatible to the C++ counterpart.

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

* [x] Add sensible notes for the reviewer
* [x] PR title is short, expressive and meaningful
* [x] Consider switching the PR to a draft (`Convert to draft`)
    * as draft PR, the CI will be skipped for pushes
* [x] Relevant issues are linked in the [References](#references) section
* [x] Every source code file has a copyright header with `SPDX-License-Identifier: Apache-2.0 OR MIT`
* [x] Branch follows the naming format (`iox2-123-introduce-posix-ipc-example`)
* [x] Commits messages are according to this [guideline][commit-guidelines]
    * [x] Commit messages have the issue ID (`[#123] Add posix ipc example`)
    * [x] Commit author matches [Eclipse Contributor Agreement][eca](and ECA is signed)
* [x] Tests follow the [best practice for testing][testing]
* [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
* [x] Assign PR to reviewer
* [x] All checks have passed (except `task-list-completed`)

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx2/blob/main/doc/release-notes/iceoryx2-unreleased.md

## Checklist for the PR Reviewer

* [x] Commits are properly organized and messages are according to the guideline
* [x] Unit tests have been written for new behavior
* [x] Public API is documented
* [x] PR title describes the changes

## Post-review Checklist for the PR Author

* [x] All open points are addressed and tracked via issues

## References

Closes #1075

### CODE REVIEW

File: iceoryx2-bb/container/src/string/mod.rs [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+
+pub use polymorphic_string::*;
+pub use relocatable_string::*;
+pub use static_string::*;
+pub use utils::*;
+
+/// Error which can occur when a [`String`] is modified.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum StringModificationError {
+    /// A string with an unsupported unicode code points greater or equal 128 (U+0080) was provided
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/src/string/mod.rs]:
```suggestion
    /// A string with unsupported unicode code points greater or equal 128 (U+0080) was provided
```

File: iceoryx2-bb/container/src/string/mod.rs [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+    use super::*;
+
+    #[doc(hidden)]
+    pub trait StringView {
+        fn data(&self) -> &[MaybeUninit<u8>];
+
+        /// # Safety
+        ///
+        /// * user must ensure that any modification keeps the initialized data contiguous
+        /// * user must update len with [`VectorView::set_len()`] when adding/removing elements
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/src/string/mod.rs]:
```suggestion
        /// * user must update len with [`StringView::set_len()`] when adding/removing elements
```

File: iceoryx2-bb/container/src/string/mod.rs [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+        unsafe fn data_mut(&mut self) -> &mut [MaybeUninit<u8>];
+
+        /// # Safety
+        ///
+        /// * user must ensure that the len defines the number of initialized contiguous
+        ///   elements in [`VectorView::data_mut()`] and [`VectorView::data()`]
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/src/string/mod.rs]:
```suggestion
        ///   elements in [`StringView::data_mut()`] and [`StringView::data()`]
```

File: iceoryx2-bb/container/src/string/mod.rs [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+        }
+
+        let pos = self.len() - bytes.len();
+        match self.rfind(bytes) {
+            Some(v) => {
+                if v != pos {
+                    return false;
+                }
+                self.remove_range(pos, bytes.len());
+                true
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/src/string/mod.rs]: `remove_range` can fail and returns a bool. Currently, it cannot fail with the check in line 374 and the `pos` calculation in line 378, but my inner chicken fears incorrect future refactoring. Can you replace these two lines with `self.remove_range(pos, bytes.len())`?

File: iceoryx2-bb/container/src/string/static_string.rs [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! Relocatable (inter-process shared memory compatible) string implementations.
+//!
+//! The [`StaticString`](crate::string::StaticString) has a fixed capacity defined at compile time.
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/src/string/static_string.rs:15]: Shall we mention the memory-layout compatibility to our C++ container here?

File: iceoryx2-bb/container/src/string/relocatable_string.rs [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! Contains the [`RelocatableString`], a
+//! run-time fixed size string that is shared memory compatible
+//!
+//! # Expert Examples
+//!
+//! ## Create [`RelocatableString`] inside constructs which provides memory
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/src/string/relocatable_string.rs]:
```suggestion
//! ## Create [`RelocatableString`] inside construct which provides memory
```

File: iceoryx2-bb/container/src/string/relocatable_string.rs [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+    #[inline(always)]
+    fn verify_init(&self, source: &str) {
+        debug_assert!(
+                self.data_ptr.is_initialized(),
+                "From: RelocatableString::{}, Undefined behavior - the object was not initialized with 'init' before.",
+                source
+            );
+    }
+
+    /// Returns the required memory size for a vec with a specified capacity
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/src/string/relocatable_string.rs]:
```suggestion
    /// Returns the required memory size for a string with a specified capacity
```

File: iceoryx2-bb/container/src/string/relocatable_string.rs [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    unsafe fn init<Allocator: iceoryx2_bb_elementary_traits::allocator::BaseAllocator>(
+        &mut self,
+        allocator: &Allocator,
+    ) -> Result<(), iceoryx2_bb_elementary_traits::allocator::AllocationError> {
+        let origin = "RelocatableString::init()";
+        if self.data_ptr.is_initialized() {
+            fatal_panic!(from origin,
+                "Memory already initialized, Initializing it twice may lead to undefined behavior.");
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/src/string/relocatable_string.rs]:
```suggestion
                "Memory already initialized. Initializing it twice may lead to undefined behavior.");
```

File: iceoryx2-bb/container/src/string/polymorphic_string.rs [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+    mem::MaybeUninit,
+    ops::{Deref, DerefMut},
+};
+
+use iceoryx2_bb_elementary_traits::allocator::{AllocationError, BaseAllocator};
+
+use crate::string::*;
+
+/// Runtime fixed-size string variant with a polymorphic allocator, meaning an
+/// allocator with a state can be attached to the vector instead of using a
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/src/string/polymorphic_string.rs]:
```suggestion
/// allocator with a state can be attached to the string instead of using a
```

File: iceoryx2-bb/container/src/string/relocatable_string.rs [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+
+impl internal::StringView for RelocatableString {
+    fn data(&self) -> &[MaybeUninit<u8>] {
+        self.verify_init("data()");
+        unsafe { core::slice::from_raw_parts(self.data_ptr.as_ptr(), self.capacity() + 1) }
+    }
+
+    unsafe fn data_mut(&mut self) -> &mut [MaybeUninit<u8>] {
+        self.verify_init("data_mut()");
+        core::slice::from_raw_parts_mut(self.data_ptr.as_mut_ptr(), self.capacity())
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/src/string/relocatable_string.rs]: Why is the passed len just `self.capacity()` instead of `self.capacity() + 1`?

File: iceoryx2-bb/container/tests/string_tests.rs [hunk 11]
```diff
@@ -0,0 +1,1066 @@
...
+        for n in 1u8..128u8 {
+            temp.push(n);
+            assert_that!(sut.push_bytes(&[n]), is_ok);
+            assert_that!(sut.len(), eq n as usize);
+            assert_that!(sut.as_bytes(), eq temp.as_slice());
+        }
+    }
+
+    #[test]
+    fn push_when_it_exceeds_the_capacity_fails<Factory: StringTestFactory>() {
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/tests/string_tests.rs]:
```suggestion
    fn push_bytes_fails_when_it_exceeds_the_capacity<Factory: StringTestFactory>() {
```

File: iceoryx2-bb/container/tests/string_tests.rs [hunk 12]
```diff
@@ -0,0 +1,1066 @@
...
+        let mut sut = factory.create_sut();
+        const RANGE_TO_FIND: [u8; 4] = [37, 38, 49, 40];
+
+        for _ in 0..(SUT_CAPACITY - 1) / 2 {
+            assert_that!(sut.push(44), is_ok);
+        }
+        assert_that!(sut.push_bytes(&RANGE_TO_FIND), is_ok);
+
+        assert_that!(sut.find(&RANGE_TO_FIND), eq Some((SUT_CAPACITY - 1)/2));
+    }
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/tests/string_tests.rs:375]: Can we also have a test for `(r)find` where the `RANGE_TO_FIND` is equal to `sut`?

File: iceoryx2-bb/container/tests/string_tests.rs [hunk 13]
```diff
@@ -0,0 +1,1066 @@
...
+    #[test]
+    fn remove_non_existing_entry_returns_none<Factory: StringTestFactory>() {
+        let factory = Factory::new();
+        let mut sut = factory.create_sut();
+
+        assert_that!(sut.remove(7), is_none);
+    }
+
+    #[test]
+    fn remove_non_existing_range_returns_false<Factory: StringTestFactory>() {
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/tests/string_tests.rs:671]: This test checks whether `remove_range` returns false when `sut.len() < idx + len`. Can you please add a test where you try to remove a non-existing range from a non-empty string?

File: iceoryx2-bb/container/tests/string_tests.rs [hunk 14]
```diff
@@ -0,0 +1,1066 @@
...
+    #[test]
+    fn truncate_to_smaller_string_works<Factory: StringTestFactory>() {
+        let factory = Factory::new();
+        let mut sut = factory.create_sut();
+        assert_that!(sut.push_bytes(b"droubadix"), is_ok);
+        sut.truncate(4);
+
+        assert_that!(sut, len 4);
+        assert_that!(sut.as_bytes_with_nul(), eq b"drou\0");
+    }
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/tests/string_tests.rs:928]: Please add a test where you try to truncate the string to `sut.len()`.

File: iceoryx2-bb/container/tests/static_string_tests.rs [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+
+#[test]
+fn from_bytes_fails_when_len_exceeds_capacity() {
+    let sut = SmallSut::from_bytes(b"oooh nooo I am toooo looong");
+    assert_that!(sut, eq Err(StringModificationError::InsertWouldExceedCapacity));
+}
+
+#[test]
+fn from_bytes_truncated_works_with_empty_bytes() {
+    let sut = Sut::from_bytes(b"");
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/tests/static_string_tests.rs]:
```suggestion
    let sut = Sut::from_bytes_truncated(b"");
```

File: iceoryx2-bb/container/tests/static_string_tests.rs [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+    assert_that!(sut, ne b"woo");
+    assert_that!(sut.as_bytes(), eq b"");
+    assert_that!(sut.as_mut_bytes(), eq b"");
+    assert_that!(sut.as_bytes_with_nul(), eq b"\0");
+    assert_that!(sut.pop(), is_none);
+}
+
+#[test]
+fn from_bytes_truncated_works_with_len_smaller_than_capacity() {
+    let sut = Sut::from_bytes(b"bonjour world");
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/tests/static_string_tests.rs]:
```suggestion
    let sut = Sut::from_bytes_truncated(b"bonjour world");
```

File: iceoryx2-bb/container/tests/static_string_tests.rs [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+
+#[test]
+fn from_str_with_len_greater_than_capacity_fails() {
+    let sut = SmallSut::from_str("the frog jumped into oblivion");
+    assert_that!(sut, eq Err(StringModificationError::InsertWouldExceedCapacity));
+}
+
+#[test]
+fn from_str_truncated_with_len_smaller_capacity_works() {
+    let mut sut = Sut::from_str("a butterfly sits on nalas nose").unwrap();
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/tests/static_string_tests.rs]:
```suggestion
    let mut sut = Sut::from_str_truncated("a butterfly sits on nalas nose").unwrap();
```

File: iceoryx2-bb/container/tests/string_utils_tests.rs [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+#[test]
+fn as_escaped_string_does_not_escape_printable_characters() {
+    for c in 32u8..128u8 {
+        let value = format!("{}", c as char);
+        assert_that!(as_escaped_string(value.as_bytes()), eq value);
+    }
+}
+
+#[test]
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/tests/string_utils_tests.rs]: This test could be moved into string_tests.rs as the `StringModificationError` is defined in mod.rs and not in utils.rs.

File: iceoryx2-bb/system-types/src/file_name.rs [hunk 19]
```diff
@@ -174,17 +174,18 @@ impl<const CAPACITY: usize> iceoryx2_bb_container::semantic_string::SemanticStri
 
     unsafe fn new_unchecked(bytes: &[u8]) -> Self {
         Self {
-            value: iceoryx2_bb_container::byte_string::FixedSizeByteString::new_unchecked(bytes),
+            value: iceoryx2_bb_container::string::StaticString::from_bytes_unchecked(bytes),
         }
     }
 
     unsafe fn insert_bytes_unchecked(&mut self, idx: usize, bytes: &[u8]) {
+        use iceoryx2_bb_container::string::String;
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/system-types/src/file_name.rs:182]: Why is this necessary?

File: iceoryx2-bb/system-types/src/file_name.rs [hunk 20]
```diff
@@ -298,6 +299,7 @@ impl<const CAPACITY: usize> core::ops::Deref for RestrictedFileName<CAPACITY> {
     type Target = [u8];
 
     fn deref(&self) -> &Self::Target {
+        use iceoryx2_bb_container::string::String;
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/system-types/src/file_name.rs:302]: Same here

File: iceoryx2-bb/container/src/string/polymorphic_string.rs [hunk 21]
```diff
@@ -0,0 +1,253 @@
...
+            data_ptr: unsafe { data_ptr.as_mut() }.as_mut_ptr().cast(),
+            len: 0,
+            capacity: capacity as _,
+            allocator,
+        })
+    }
+
+    /// Same as clone but it can fail when the required memory could not be
+    /// allocated from the [`BaseAllocator`].
+    pub fn try_clone(&self) -> Result<Self, AllocationError> {
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx2-bb/container/src/string/polymorphic_string.rs:217]: This is not tested

**Review by FerdinandSpitzschnueffler — COMMENTED** [review]
Please update the release notes.

File: iceoryx2-bb/system-types/src/file_name.rs (hunk 19 above)
Comment (elfenpiff) [reply iceoryx2-bb/system-types/src/file_name.rs:182]: because I need the trait in scope to access the methods of the trait. and in the file I require the std `String` - both have the same name and clash - either I put the std::string::String in globally or the trait - with this approach it works.

**Review by FerdinandSpitzschnueffler — COMMENTED** [review]
LGTM. Can be merged when the release notes are updated and the CI is green.

**Comment by elfenpiff** [comment]
I updated the release notes in the follow up PR: https://github.com/ekxide/iceoryx2/commit/1b9dc4e1923ce87e61a4f3e40b525278f1aa5119 to safe some CI time and get this merged.
